# RSA Algorithm

In [1]:
#importing necessary packages
import random
import time

Using Euclid's Algorithm for finding the greatest common divisor: 

In [2]:
def gcd(a,b):
    while b!=0:
        a,b=b,a%b
    return a

Function to calculate d using multiplicative inverse:

In [3]:
def multiplicative_inverse(e, phi):
    d = 0
    x1 = 0 
    x2 = 1
    y1 = 1
    temp_phi = phi
    
    while e > 0:
        temp1 = temp_phi//e
        temp2 = temp_phi - temp1 * e 
        temp_phi = e
        e = temp2
        
        x = x2-temp1 * x1 
        y = d-temp1 * y1
        
        x2 = x1
        x1 = x
        d = y1
        y1 = y
    
    if temp_phi == 1: 
        return d + phi

Function to check for prime number:

In [4]:
def is_prime (num):
    if num == 2: 
        return True
    if num < 2 or num % 2 == 0:
        return False
    for n in range (3, int(num**0.5)+2, 2):
        if num % n == 0: 
            return False
    return True

Function to generate the public and private key pairs

In [5]:
def generate_key_pair(p, q): 
    if not (is_prime(p) and is_prime(q)): 
        raise ValueError('Both numbers must be prime')
    elif p == q:
        raise ValueError('p and q cannot be equal')
    
    n = p*q
    phi = (p-1)*(q-1)
    e = random.randrange(1,phi)
    g = gcd(e,phi)
                         
    while g!=1:
        e = random.randrange(1,phi)
        g = gcd(e, phi)
    
    d=multiplicative_inverse(e,phi)

    print("e: ", e)
    print("n: ", n) 
    print("d: ", d)      
    
    # Return public and private key_pair 
    # Public key is (e, n) and private key is (d, n) 
    return ((e, n), (d, n))

Define encryption function

In [6]:
def encrypt (pk, plaintext):
    key, n = pk
    cipher= [pow(ord(char), key, n) for char in plaintext]
    return cipher

Rsa Encryption main method

In [7]:
if __name__ == "__main__":
    print("=================================RSA ENCRYPTION=======================================")
    
    p = int (input("Enter a prime number (17, 19, 23, etc): "))
    q = int(input("Enter another prime number (Different from the one entered above): "))
    
    print("Generating public / private key-pairs now ...")
    public, private = generate_key_pair (p, q)
    print("The public key is ", public," and the private key is ", private)
    
    message = input(" Enter a message to encrypt with the public key: ")
    encrypted_msg = encrypt (public, message)
    print("The encrypted message is: ", ''.join(map (lambda x: str(x), encrypted_msg)))

=================================RSA ENCRYPTION=======================================
Enter a prime number (17, 19, 23, etc): 223
Enter another prime number (Different from the one entered above): 379
Generating public / private key-pairs now ...
e:  20725
n:  84517
d:  59881
The public key is  (20725, 84517)  and the private key is  (59881, 84517)
 Enter a message to encrypt with the public key: anirban
The encrypted message is:  43188221022565948987615214318822102


Now Decrypting the message:

In [8]:
start=time.time()
def decrypt(pk, ciphertext):
    key,n=pk
    aux=[str(pow(char,key,n)) for char in ciphertext]
    plain=[chr(int(char2)) for char2 in aux]
    return ''.join(plain)

print("Decrypting message with private key",private,"...")
print("The Message is:",decrypt(private,encrypted_msg))
print(" ")
print('Time taken=',time.time()-start,'seconds')

Decrypting message with private key (59881, 84517) ...
The Message is: anirban
 
Time taken= 0.00099945068359375 seconds


# Using Brute Force

Let's try cracking the same key (d value), using Brute force:

In [10]:
import time
start = time.time()

def egcd(a, b):
    if a == 0:
        return (b, 0, 1)
    else:
        g, y, x = egcd(b%a,a) 
        return (g, x-(b//a)*y, y)
    
def modinv (a, m):
    g, x, y = egcd (a, m)
    if g != 1:
        raise Exception ('modular inverse does not exist')
    else:
        return x%m
    

def factor (n):
    for i in range (3, n):
        if n%i == 0:
            return i

e = 20725
n = 84517

p = factor(n)
q = n//p 
phi_n=(p-1)*(q-1)

d_crack = modinv(e, phi_n)
print("Cracking value of d using Brute force...") 
print('Cracked value of d:', d_crack)
print('Time taken to crack using brute force:', time.time ()-start, 'seconds.')

Cracking value of d using Brute force...
Cracked value of d: 59881
Time taken to crack using brute force: 0.0010001659393310547 seconds.


We see that the difference between Brute force attack and Algorithmic decryption (0.0010001659393310547 - 0.00099945068359375) is 0.00000071525 seconds.